96.md

```diff
@@ -0,0 +1,221 @@
+NIP-96
+======
+
+HTTP File Storage Integration
+-----------------------------
+
+`draft` `optional` `author:arthurfranca` `author:Semisol` `author:staab`
+
+## Introduction
+
+This NIP defines a REST API for HTTP file storage servers intended to be used in conjunction with the nostr network.
+The API will enable nostr users to upload files and later reference them by url on nostr notes.
+
+The spec DOES NOT use regular nostr events through websockets for
+storing, requesting nor retrieving data because, for simplicity, the server
+will not have to learn anything about nostr relays.
+
+## Server Adaptation
+
+File storage servers wishing to be accessible by nostr users should opt-in by making available an https route at `/.well-known/nostr/nip96.json` with `api_url`:
+
+```js
+{
+  // Required
+  // File upload and deletion are served from this url
+  // Also downloads if "download_url" field is absent or empty string
+  "api_url": "https://your-file-server.example/custom-api-path",
+  // Optional
+  // If absent, downloads are served from the api_url
```

Comment:
"api_url" field was just to merge "delete" and "upload", depending if it is POST or DELETE action.

At this moment this NIP is just for basic functionallity (upload/delete/download one file at a time). Your server can have extra features outside of the NIP scope or we can expand it after PR merge. 

Do you need me to split delete and upload to their own urls for some reason?

Reply:
I think I got what you mean. We do need atleast an url to search for files by hash (so one can search at many servers till one finds it).

But what if we keep the download_url at `/.well-known/nostr/nip96.json` so that when trying to find a file at other servers, I can use this fixed one (or api_url when absent). **BUT** what if after uploading a file, the upload response includes an url to download the uploaded file? Now you can kind of load balance by sending different urls at each upload like:

Upload 1: response has url "https://your-file-server.example/path-A/hash1"
Upload 2: response has url "https://your-file-server.example/path-B/hash2"
Upload 3: response has url "https://your-file-server.example/path-C/hash3"
Upload 4: response has url "https://your-file-server.example/path-A/hash1"
and so on..

In the example you have a group of 3 different download routes you use so effectively load balancing, because the nostr events will use them instead of what is informed at the config file.

Reply:
Alright. You can do it here at the "url" upload response field:

https://github.com/nostr-protocol/nips/blob/581340338d93de40e8801f3eba689420b6832305/96.md?plain=1#L147-L150

It can be any hostname, any path, any filename. This url will be the one used inside nostr events to download the file.

But, for decentralization purposes, you still have to support this fixed url when the user just knows the original file hash and wanna try downloading from you:

https://github.com/nostr-protocol/nips/blob/581340338d93de40e8801f3eba689420b6832305/96.md?plain=1#L223-L225


Does this solve the problem?